File: app/src/main/java/com/example/architectureapplication/di/AppComponent.java
package com.example.architectureapplication.di;

import android.app.Application;

import javax.inject.Singleton;

import dagger.BindsInstance;
import dagger.Component;

@Singleton
@Component(modules = {AppModule.class, AppSubComponents.class})
public interface AppComponent {

    @Component.Builder
    interface Builder{

        @BindsInstance
        Builder application(Application application);

        AppComponent build();
    }


    MainComponent.Builder mainComponent();
    AddNoteComponent.Builder addNoteComponent();
}

File: app/src/main/java/com/example/architectureapplication/kt/AppComponent.kt
package com.example.architectureapplication.kt

import android.app.Application
import dagger.BindsInstance
import dagger.Component
import javax.inject.Singleton

@Singleton
@Component(modules = arrayOf(AppModule::class))
interface AppComponent {

    fun mainComponent(): MainComponent.Builder
    fun addNoteComponent(): AddNoteComponent.Builder

    @Component.Builder
    interface Builder{

        @BindsInstance
        fun application(app:Application): Builder

        fun build():AppComponent
    }
}
